_posts/2020-01-30-database-migration-spring-boot-flyway.md

```diff
@@ -0,0 +1,134 @@
+
+We use H2 database in an `in-memory` mode for this post. Spring Boot auto-configures [Flyway](https://docs.spring.io/spring-boot/docs/current/reference/html/howto.html#howto-execute-flyway-database-migrations-on-startup) and [H2](https://docs.spring.io/spring-boot/docs/current/reference/html/spring-boot-features.html#boot-features-embedded-database-support) as long as we add the following dependencies to our build file (Gradle notation):
+
+```
+implementation 'org.flywaydb:flyway-core'
+runtimeOnly 'com.h2database:h2:1.4.199'
+```
+
+This is enough for Spring Boot to initializes a specific implementation of `DataSource` called `EmbeddedDatabase`. It is  then used by Flyway to execute the given migrations. All configuration options can be found at [Spring Boot application properties reference](https://docs.spring.io/spring-boot/docs/current/reference/html/appendix-application-properties.html)
+
+In case we prefer to use one of the build tool plugins, an additional definition is required in our build file (Gradle notation):
+
+```gradle
+plugins {
+
+  // Other plugins...
+ 
+  id "org.flywaydb.flyway" version "6.2.0"
+}
+
+flyway {
+  url = 'jdbc:h2:mem:'
+}
+```
+
+## Writing Our First Database Migration
+
+Flyway has own [naming convention](https://flywaydb.org/documentation/migrations#naming) which can be adjusted to our needs using the following [configurations](https://docs.spring.io/spring-boot/docs/current/reference/html/appendix-application-properties.html#data-migration-properties):
+
+```
+spring.flyway.sql-migration-prefix
+spring.flyway.sql-migration-separator
+spring.flyway.sql-migration-suffixes
+```
+
+Here is the `V1__init.sql` file used in our [code examples repository](https://github.com/thombergs/code-examples/tree/master/spring-boot) **TODO: Put actual link here when the code is approved**:
+
+```sql
+CREATE TABLE auth_user(
+    id INT AUTO_INCREMENT PRIMARY KEY,
+    username VARCHAR(255) NOT NULL unique
+);
+```
+
+## Running Flyway
+
+Choosing the right running option depends on our needs and Flyway tries to cover almost all of them - [command-line](https://flywaydb.org/documentation/commandline/), [Java/Android API](https://flywaydb.org/documentation/api/), [Maven](https://flywaydb.org/documentation/maven/)/[Gradle](https://flywaydb.org/documentation/gradle/) plugins, [Docker](https://hub.docker.com/r/flyway/flyway) and a decent list of [community plugins and integrations](https://flywaydb.org/documentation/plugins/).
+
+By default, Spring Boot runs Flyway database migrations on [startup](https://docs.spring.io/spring-boot/docs/current/reference/html/howto.html#howto-execute-flyway-database-migrations-on-startup) by looking at `classpath:db/migration` folder. We can modify this location by passing a comma-separated list of one or more `classpath:` or `filesystem:` locations to `spring.flyway.locations` property.
+
+When the Gradle plugin is preferred, we can call the following command in our terminal: 
```

Comment:
Isn't the information provided in `Setting up Flyway` section enough? If not please give me some hints what would you like to see.

Reply:
I would pull all information about the Gradle Plugin into its own section (or sections, if it becomes too long). How to set it up, and then how to use it, and what its pros/cons are. In the Setting up Flyway section, just assume the default option on startup. This way, we have each option cleanly separated into its own section and the reader doesn't have to jump between sections to follow a certain option.

I'm missing an explanation how to configure the Gradle plugin. How does it know which database to connect to? What is the `-i` doing? What is the command `flywayMigrate` and which other commands are possible?

Reply:
I even think that it would be better if I combine `Setting up` and `Running` sections so that we have the following structure:
## Setting up and Running Flyway
### Spring Boot auto-configuration
### Gradle plugin
.....
In this way we will have configuration and execution for each option in one place. What do you think?

Reply:
Yes, this makes it very clear, I like that!